src/link/inaccessible_spec_definition.rs

```diff
@@ -0,0 +1,1137 @@
+    }
+}
+impl IsInaccessibleExt for position::EnumValueDefinitionPosition {
+    fn is_inaccessible(
+        &self,
+        schema: &FederationSchema,
+        inaccessible_directive: &Name,
+    ) -> Result<bool, FederationError> {
+        // NOTE It'd be more efficient to start at parent and look up the value directly from there.
+        let value = self.get(schema.schema())?;
+        Ok(value.directives.has(inaccessible_directive)
+            || self
+                .parent()
+                .is_inaccessible(schema, inaccessible_directive)?)
+    }
+}
+impl IsInaccessibleExt for position::DirectiveArgumentDefinitionPosition {
+    fn is_inaccessible(
+        &self,
+        schema: &FederationSchema,
+        inaccessible_directive: &Name,
+    ) -> Result<bool, FederationError> {
+        let argument = self.get(schema.schema())?;
+        Ok(argument.directives.has(inaccessible_directive))
+    }
+}
+
+/// Types can be referenced by other schema elements in a few ways:
+/// 1. Fields, arguments, and input fields may have the type as their base
+///    type.
+/// 2. Union types may have the type as a member (for object types).
+/// 3. Object and interface types may implement the type (for interface
+///    types).
+/// 4. Schemas may have the type as a root operation type (for object
+///    types).
+///
+/// When a type is hidden, the referencer must follow certain rules for the
+/// schema to be valid. Respectively, these rules are:
+/// 1. The field/argument/input field must not be in the API schema.
+/// 2. The union type, if empty, must not be in the API schema.
+/// 3. No rules are imposed in this case.
+/// 4. The root operation type must not be the query type.
+///
+/// This function validates rules 1 and 4.
+fn validate_inaccessible_type(
+    schema: &FederationSchema,
+    inaccessible_directive: &Name,
+    position: &TypeDefinitionPosition,
+    errors: &mut MultipleFederationErrors,
+) -> Result<(), FederationError> {
+    let referencers = schema.referencers();
+
+    macro_rules! check_inaccessible_reference {
+        ( $ty:expr, $ref:expr ) => {
+            if !$ref.is_inaccessible(schema, inaccessible_directive)? {
+                errors.push(SingleFederationError::ReferencedInaccessible {
+                    message: format!("Type `{}` is @inaccessible but is referenced by `{}`, which is in the API schema.", $ty, $ref),
+                }.into())
+            }
+        }
+    }
+
+    macro_rules! check_inaccessible_referencers {
+        ( $ty:expr, $( $referencers:expr ),+ ) => {
+            $(
+                for ref_position in $referencers {
+                    check_inaccessible_reference!(position, ref_position);
+                }
+            )+
+        }
+    }
+
+    match position {
+        TypeDefinitionPosition::Scalar(scalar_position) => {
+            let referencers = referencers
+                .scalar_types
+                .get(&scalar_position.type_name)
+                .unwrap();
```

Comment:
I guess it should never be the case that this would panic but just wondering whether we should be more defensive here just in case

Reply:
Yeah this would only happen if there is a bug, so I don't think panicking is _wrong_, but it's different from how we treat this case elsewhere, so I can change it to match the rest